### ISSUE TITLE
Spurious "Error: upper bound above length of sequence" when slicing return-value sequence

### ISSUE THREAD

**Issue by brady-ds** [dafny-lang/dafny#202]
On

````
method push<T>(sequence: seq<T>, element: T) returns (new_sequence: seq<T>)
  ensures sequence == new_sequence[..|new_sequence|-1]
{
  return sequence + [element];
}
````

Dafny 2.1.1.10209 reports the spurious error

````
Dafny 2.1.1.10209
push.dfy(2,34): Error: upper bound above length of sequence
Execution trace:
    (0,0): anon0

Dafny program verifier finished with 1 verified, 1 error
````

When `new_sequence` is not a return value, the error disappears:

````
method no_problem<T>(sequence: seq<T>, element: T) {
  var new_sequence := sequence + [element];
  assert sequence == new_sequence[..|new_sequence|-1];
}
````

**Comment by wilcoxjay** [comment]
The problem here is that when checking the well-formedness of the postcondition, Dafny treats the return value as "opaque". In other words, it assumes nothing other than type-correctness. In your example, that means that Dafny considers the possibility that `new_sequence` might be empty (even though this can never happen), in which case the expression `new_sequence[..|new_sequence|-1]` is ill formed, since the index is negative.

The fix to your program is to add the postcondition `|new_sequence| > 0` before the original postcondition. Then the program will verify.

There is a second problem you have identified, which is that the error message Dafny reports here is confusing. It claims that the index might be too big, when in fact it might be too small. I am testing a fix for this locally and will push it soon.

**Comment by brady-ds** [comment]
Thanks!  Yes, please consider this a bug about the error message, as with a better one I would have seen the fix.

**Comment by davidcok** [comment]
Confirmed that this is a duplicate of #326. Adding a test case along with the fix to #326.
